torchao/quantization/linear_activation_scale.py

```diff
@@ -0,0 +1,165 @@
+from typing import Any, Callable, Dict, Optional
+
+import torch
+from torch.utils._python_dispatch import return_and_correct_aliasing
+
+from torchao.utils import (
+    TORCH_VERSION_AT_LEAST_2_5,
+    TorchAOBaseTensor,
+)
+
+__all__ = [
+    "WeightTensorWithLinearActivationScaleMetadata",
+    "to_weight_tensor_with_linear_activation_scale_metadata",
+]
+
+aten = torch.ops.aten
+
+
+class WeightTensorWithLinearActivationScaleMetadata(TorchAOBaseTensor):
+    """
+    Tensor subclass that wraps a weight tensor and provides metadata for linear activation scaling.
+    Right now we hardcode how we apply the scale:
+       scaled_linear_act = input_act / scale
+       out = F.linear(scaled_linear_act, weight, ...)
+
+    We can generalize this to accept a function as well if needed.
+
+    Args:
+        original_weight_tensor (torch.Tensor): The weight tensor to be wrapped.
+        scale (torch.Tensor): The scale tensor to be applied to activation.
+    """
+
+    original_weight_tensor: torch.Tensor
+    scale: torch.Tensor
+
+    def __new__(
+        cls,
+        original_weight_tensor: torch.Tensor,
+        scale: torch.Tensor,
+    ):
+        kwargs = {}
+        dtype = original_weight_tensor.dtype
+        kwargs["dtype"] = dtype
+        kwargs["requires_grad"] = False
+        kwargs["device"] = original_weight_tensor.device
+        shape = original_weight_tensor.shape
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        original_weight_tensor: torch.Tensor,
+        scale: torch.Tensor,
+    ):
+        self.original_weight_tensor = original_weight_tensor
+        self.scale = scale
+
+    def __repr__(self):
+        return f"WeightTensorWithLinearActivationScaleMetadata({self.original_weight_tensor}, scale={self.scale}"
+
+    def __tensor_flatten__(self):
+        tensor_data = ["original_weight_tensor", "scale"]
+        return tensor_data, []
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        return cls(
+            tensor_data_dict["original_weight_tensor"],
+            tensor_data_dict["scale"],
+        )
+
+    @staticmethod
+    def _quantized_linear_op(
+        input_tensor: torch.Tensor, weight_tensor: torch.Tensor, bias: torch.Tensor
+    ):
+        original_weight_tensor = weight_tensor.original_weight_tensor
```

Comment:
In the example a quantized tensor is passed in so instead of being called `original_weight_tensor` shouldn't it be called `quantized_weight_tensor`

Reply:
the tensor subclass works with either quantized or not quantized tensor actually

`original_weight_tensor` means the original weight tensor that is not wrapped with this tensor subclass (in our case it's a quantized tensor)